src/map/utils/battleutils.cpp

```diff
@@ -1825,6 +1826,23 @@ namespace battleutils
                 return base + skillModifier;
             }
         }
+        else if (PDefender->objtype == TYPE_TRUST)
+        {
+            CMobEntity* PTrust = (CMobEntity*)PDefender;
+            if (PTrust->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
+            {
+                base = PDefender->getMod(Mod::SHIELDBLOCKRATE);
+                if (PDefender->StatusEffectContainer->HasStatusEffect(EFFECT_REPRISAL))
+                {
+                    base = base * 1.5f;
```

Comment:
I don't think checking Reprisal and returning base block rate is needed, reprisal and the combat skill check for block rate is calculated normally, otherwise you'd have to copy paste in all the missing stuff

reducing it to 

```
            if (PTrust->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
            {
                base = PDefender->getMod(Mod::SHIELDBLOCKRATE);
            }
```
should accomplish everything we need, which was to establish base block rate.... but that would get `SHIELDBLOCKRATE` added a second time, boosting it higher than it should. @TeoTwawki thoughts? new mobmod/mod for base block rate or copy paste in the skill calculations?

This block would add back in the skill calc:
```
            if (PTrust->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
            {
                base = PDefender->getMod(Mod::SHIELDBLOCKRATE);
                
                // Check for Reprisal and adjust skill and block rate bonus multiplier
                if (PDefender->StatusEffectContainer->HasStatusEffect(EFFECT_REPRISAL))
                {
                    blockSkill   = blockSkill * 1.15f;
                    reprisalMult = 1.5f; // Default is 1.5x

                    // Adamas and Priwen set the multiplier to 3.0x while equipped
                    if (PDefender->getMod(Mod::REPRISAL_BLOCK_BONUS) > 0)
                    {
                        reprisalMult = 3.0f;
                    }
                }
                
                skillModifier = (blockSkill - attackSkill) * 0.2325f;

                // Add skill and Palisade bonuses
                base += skillModifier + palisadeMod;
                // Multiply by Reprisal's bonus
                base = base * reprisalMult;

                // Apply the lower and upper caps
                blockRate = (base < 5) ? 5 : base;
                blockRate = (base > 100) ? 100 : base;
                
                return blockRate
            }
```

which is leaving in REPRISAL_BLOCK_BONUS check in case someone wanted custom content with Priwen/Adamas trust or something

Reply:
I was thinking, there are mobs that should also be able to block and I am 99% sure they don't right now.

I thought we we could do something like:
```cpp
if (PDefender->objtype != TYPE_PC)
{
    if (PDefender->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
    {
        base = PDefender->getMod(Mod::SHIELDBLOCKRATE);

        if (PDefender->objtype == TYPE_PET)
        {
            // Automaton specific skillcheck
        }
        else
        {
            // trust and mob specific skillcheck 
            // (it's the same code of checking its skillcap for its job's rank regardless which it is)
        }
    }
}
```
SHIELDBLOCKRATE is just standing in for the shield-size they don't actually have (despite appearances on the model).

Using the skill is the already established pattern for extant code. (I don't know if that was right, I just know it's what was there before we started - it's hard to tell if automaton blocks were ever right or wrong when I've never played it)

Reprisal and optional custom bonuses would then do its thing after the skill checking per your description above.

Reply:
I can adjust this, but CBattleEntity wouldn't set getMobMob, so I had to reassign PDefender to CMobEntity.
```
CMobEntity* PTrust = (CMobEntity*)PDefender;
if (PTrust->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
```

Reply:
above was kinda pseudo code, but we'd start off just converting the pet block over to a "non player" block with the other entity checks within it, and the skill-calc for mob and trust would be identical, entity casting aside was what I was getting at.

Reply:
seems reasonable to me

Reply:
the mobMod will need to be added to the Automatons also?

Reply:
> the mobMod will need to be added to the Automatons also?

I was thinking it'd be added to all non-players we want to do blocks. So where valoredge currently is getting its "hey you are a tank now bro" we'd give it the mobmod. Looks like that would be over here, and [gosh look at that](https://github.com/LandSandBoat/server/blob/ff257cf51c312c5b186e203410318b066ccf3f1c/src/map/utils/petutils.cpp#L712), it really is a "phantom shield". I hate this so much. Non Players don't have gear but we're cramming a value into an imaginary slot just to have it in memory. _and then proceeded to not use the "size" set there and hardcode a value **anyway**_   Thanks past devs, I hate it.

Reply:
```
        if (PDefender->objtype == TYPE_PC)
        {
            CCharEntity*    PChar = (CCharEntity*)PDefender;
            CItemEquipment* PItem = PChar->getEquip(SLOT_SUB);

            if (PItem)
            {
                shieldSize = PItem->getShieldSize();
            }
            else
            {
                return 0;
            }
        }
        else if (PDefender->objtype != TYPE_PC)
        {
            CMobEntity* PEntity = (CMobEntity*)PDefender;
            if (PEntity->getMobMod(MOBMOD_CAN_SHIELD_BLOCK) > 0)
            {
                base = PDefender->getMod(Mod::SHIELDBLOCKRATE);
                if (PDefender->objtype == TYPE_PET)
                {
                    skillModifier = (int8)((PDefender->GetSkill(SKILL_AUTOMATON_MELEE) - attackSkill) * 0.215f);
                    if (base <= 0)
                    {
                        return 0;
                    }
                    else
                    {
                        return base + skillModifier;
                    }
                }
                else if (PDefender->objtype == TYPE_TRUST)
                {
                    // Check for Reprisal and adjust skill and block rate bonus multiplier
                    if (PDefender->StatusEffectContainer->HasStatusEffect(EFFECT_REPRISAL))
                    {
                        blockSkill   = blockSkill * 1.15f;
                        reprisalMult = 1.5f; // Default is 1.5x

                        // Adamas and Priwen set the multiplier to 3.0x while equipped
                        if (PDefender->getMod(Mod::REPRISAL_BLOCK_BONUS) > 0)
                        {
                            reprisalMult = 3.0f;
                        }
                    }

                    skillModifier = (blockSkill - attackSkill) * 0.2325f;

                    // Add skill and Palisade bonuses
                    base += skillModifier + palisadeMod;
                    // Multiply by Reprisal's bonus
                    base = base * reprisalMult;

                    // Apply the lower and upper caps
                    blockRate = (base < 5) ? 5 : base;
                    blockRate = (base > 100) ? 100 : base;

                    if (PDefender->objtype == TYPE_TRUST)
                    {
                        printf("battleutils.cpp GetBlockRate TRUST: [%s] BLOCKRATE: [%3.0f]\n", PDefender->GetName(), blockRate);
                    }

                    return blockRate;
                }
                else if (PDefender->objtype == TYPE_MOB)
                {
                    if (PMob->m_EcoSystem != ECOSYSTEM::UNDEAD && PMob->m_EcoSystem != ECOSYSTEM::BEASTMAN)
                    {
                        return 0;
                    }
                }
            }
        }
```
Mob would have same calc as trusts? do they use reprisal and palisade? (honestly don't know)

Reply:
It's not so much that they "do" but they "could" so we don't have to rework it later

Reply:
ahh, got it. 👍

Reply:
Change this `else if (PDefender->objtype == TYPE_TRUST)` to just an `else`

and then

Delete this block, as it's no longer necessary
```                
                else if (PDefender->objtype == TYPE_MOB)
                {
                    if (PMob->m_EcoSystem != ECOSYSTEM::UNDEAD && PMob->m_EcoSystem != ECOSYSTEM::BEASTMAN)
                    {
                        return 0;
                    }
                }
```

This way we can add mobmods to allow things to block when they should be able to

For automatons we can remove `PPet->m_Weapons[SLOT_SUB]->setShieldSize(3);` and `setMobMod()` that mod that was created in this PR. Don't know if that shieldSize is even used at all but we need to set the mod anyway

Reply:
@TeoTwawki @WinterSolstice8 

Blockskill pulls Skill_Shield for calculations. Will need to add that back into trustutils for the blocking calculations. ![image](https://user-images.githubusercontent.com/99501957/198306485-05cc4032-b1d3-4a01-8e45-e63d29e65d01.jpeg)

Reply:
> @TeoTwawki @WinterSolstice8
> 
> Blockskill pulls Skill_Shield for calculations. Will need to add that back into trustutils for the blocking calculations. !

The calculating can be done anywhere because we don't need to get skill they actually "have" like a player would, instead we're getting what the cap _would be_ if they _were_ a player with maxed out skill for the level, make sense?

Reply:
Adding SKILL_SHIELD for trust

![image](https://user-images.githubusercontent.com/99501957/198310778-e391ecc0-d4de-42e9-a87c-7e705fa20c0b.png)

Adding base block rate to valor edge 

![image](https://user-images.githubusercontent.com/99501957/198317773-af1da14d-c350-4257-a814-dcc3243a8695.png)

Reply:
> > @TeoTwawki @WinterSolstice8
> 
> > 
> 
> > Blockskill pulls Skill_Shield for calculations. Will need to add that back into trustutils for the blocking calculations. !
> 
> 
> 
> The calculating can be done anywhere because we don't need to get skill they actually "have" like a player would, instead we're getting what the cap _would be_ if they _were_ a player with maxed out skill for the level, make sense?
> 
> 
> 
> we fake skills for mobs all over the core this way for various things.
> 
> 

Wellllll I just submitted it right as you commented 🙄 it was flooring rates to -96 with the blockskill pulling a 0 into the calculation. Adding the skill hit the multiplier like a player and can still be cranked to cap with the mod if an admin so chooses.

Reply:
what we do on mobs is instead of checking skill of the entity, we check skill of its job(s) at that level:

This example is checking a subjob (because it was a fallback for mobs that cast spells from a sub when main has no magic - it was just 1st example I found)
```cpp
battleutils::GetMaxSkill((SKILLTYPE)i, PMob->GetSJob(), mLvl > 99 ? 99 : mLvl); instead of checking skill on the actual entity, with mobs we check the skill of their job's cap at that level
```

Reply:
> Wellllll I just submitted it right as you commented roll_eyes it was flooring rates to -96 with the blockskill pulling a 0 into the calculation. Adding the skill hit the multiplier like a player and can still be cranked to cap with the mod if an admin so chooses.

I just woke up and checked my notices so I could not catch you sooner, sorry

Reply:
> what we do on mobs is instead of checking skill of the entity, we check skill of its job(s) at that level:
> 
> ```c++
> battleutils::GetMaxSkill((SKILLTYPE)i, PMob->GetSJob(), mLvl > 99 ? 99 : mLvl); 
> ```
> 
> This example is checking a subjob because it was a fallback for mobs that cast spells from a sub when main has no magic - it was just 1st example I found
thats what i thought this was?
`battleutils::GetMaxSkill(SKILLTYPE::SKILL_SHIELD, mJob, mLvl > 99 ? 99 : mLvl);
            if (maxSkill != 0)`

Reply:
if it helps you understand me, I can't actually see your code anymore, only your comments/notifications that describe what you said you did. github here is totally broken for me on this pr. anytime I you to look at the file changes the page goes "not responding".

So I am relying on what you said you did and winter's suggestions that were among them before I got here.

I saw you say you were giving them skill in trust utils and made the logical conclusions. And the setting is still bunk I tell you, bunk! _sulks_

Reply:
Should I close and reopen this as a new PR how it stands for clarity? I really want to do this right for continuity.